python/lsst/daf/butler/_butler.py

```diff
@@ -1869,16 +1873,56 @@ def transfer_from(self, source_butler: Butler, source_refs: Iterable[DatasetRef]
 
         # Importing requires that we group the refs by dataset type and run
         # before doing the import.
+        source_dataset_types = set()
         grouped_refs = defaultdict(list)
         grouped_indices = defaultdict(list)
         for i, ref in enumerate(source_refs):
             grouped_refs[ref.datasetType, ref.run].append(ref)
             grouped_indices[ref.datasetType, ref.run].append(i)
-
-        # Register any dataset types we need. This has to be done outside
-        # of a transaction and so will not be rolled back on failure.
-        for datasetType, _ in grouped_refs:
-            self.registry.registerDatasetType(datasetType)
+            source_dataset_types.add(ref.datasetType)
+
+        # Check to see if the dataset type in the source butler has
+        # the same definition in the target butler and register missing
+        # ones if requested. Registration must happen outside a transaction.
+        missing_dataset_types = set()
+        inconsistent_dataset_types = {}
+        newly_registered_dataset_types = set()
+        for datasetType in source_dataset_types:
+            is_missing = False
+            try:
+                # If another process is registering this just after this
+                # search then this process could report failure if not
+                # allowed to register. That is fine. Rerunning will fix
+                # the problem.
+                target_dataset_type = self.registry.getDatasetType(datasetType.name)
```

Comment:
Now that I look at it again, I think that if I just used `registerDatasetType` then that code would only trigger if we are registering dataset types. This would mean that the first we might know about a dataset type not being registered or being inconsistent might be half an hour later when we try to import some data with that dataset type and then rollback the entire transaction. Don't we want to have a hint up front?

Reply:
`registerDatasetType` does return a `bool` already.

> This would mean that the first we might know about a dataset type not being registered or being inconsistent might be half an hour later when we try to import some data with that dataset type and then rollback the entire transaction. Don't we want to have a hint up front?

I was thinking that even a half hour earlier is too late, at least for execution butler - we need to get that warning when we make the QG, as per DM-30845.  But yeah, more generally, having an earlier check does seem like a good idea.  How about doing this up front; this still minimizes DB calls and maximizes reuse of checks in daf_butler.
```
if register_dataset_types:
    if butler.registry.registerDatasetType(datasetType):  # let this raise if inconsistent
        log.verbose(...)
else:
    if butler.registry.getDatasetType(name) != datasetType):
        raise ConflictingDefinitionError(...)  # what registerDatasetTypes raises
```
?

Makes me wonder if we should put the else block into an `Registry.expectDatasetType` or an `expect=True` option to `registerDatasetType`, as I bet this isn't the only place that wants this logic.

Reply:
I rewrote as described.